technology-considerations/s3-user-access-to-objects.md

```diff
@@ -0,0 +1,121 @@
+# Restricting User Access To S3 Objects
+
+## Problem
+
+Ensure that TDR users are not able to overwrite/view another user’s objects in the TDR AWS S3 upload bucket.
+
+Potential options to resolve the problem are to use:
+* AWS Cognito;
+* Signed URLS.
+
+The purpose of the spike was to see whether using AWS Cognito was a feasible option.
+
+AWS Cognito has some advantages over using signed URLs:
+* Using signed URLs means an eavesdropper, potentially, can grab the signed url and use it until it expires
+* AWS Cognito will work with AWS KMS encrypted AWS S3, can set up an AWS IAM or key policy allowing access to the AWS KMS key from AWS Cognito identities
+
+### Additional Considerations
+
+Unclear how AWS Cognito will work with Javascript silent login using Keycloak's Javascript library as the tokens are stored in Redis and not cookies.
+
+## Using AWS Cognito and Keycloak to Restrict Access to AWS S3 Bucket Objects
+
+### AWS IAM Policy To Restrict Access to AWS S3 Bucket Objects
+
+The following AWS IAM policy is an example of using an AWS Cognito identity sub parameter to control access to AWS S3 bucket objects.
+
+Note: The `cognito-identity.amazonaws.com:sub` refers to the identity id associated with the user in the AWS Cognito Identity Pool.
+
+```
+    {
+        "Version": "2012-10-17", 
+        "Statement": [
+            {
+                "Sid": "ListYourObjects",
+                "Effect": "Allow",
+                "Action": "s3:ListBucket",
```

Comment:
The policy we'll use will just be `s3:PutObject`, or whatever we need to let users upload files. We don't want to add any read permissions, because there's no plan in the roadmap to let users re-download their files through TDR.

The authorisation is exactly the same, but it might be worth updating this to `PutObject` so we don't get confused later.

Reply:
Adjusted policy